profiles/coreos/base/profile.bashrc.slsa-provenance

```diff
@@ -0,0 +1,182 @@
+# Build provenance hooks
+# ======================
+# The functions below hook into every ebuild's execution and generate provenance files
+#   to meet the SLSA provenance requirements (https://slsa.dev/spec/v0.1/requirements#available).
+# All input files (source tarball / git commit hash) plus added patches / files,
+#   and all resulting installation binaries and files are captured.
+# The information is emitted in SLSA provenance 0.2 format (see https://slsa.dev/provenance/v0.2)
+
+
+# We only record provenance when a package is actually being built.
+# See profiles/coreos/base/profile.bashrc for cros_...
+cros_post_src_configure_enable_slsa_provenance_report() {
+    if [ "${GENERATE_SLSA_PROVENANCE:-}" != "true" ] ; then
+        einfo "Provenance generation not requested by build; skipping."
+        return 0
+    fi
+    export __slsa_provenance_report="yes"
+}
+# --
+
+# Generate SLSA provenance 0.2 Subject information.
+# The information will cover all installation files shipped with a package.
+__slsa_provenance_subject() {
+    local parallel="$(grep -cE '^processor' /proc/cpuinfo)"
+    local comma=""
+
+    einfo "    Provenance: recording subject (output)"
+
+    echo '   "subject": ['
+
+    (
+        cd "$D"
+        find . -type f -print | sed 's:^./::' | xargs -P "$parallel" -L 1 sha512sum
+    ) | while read checksum filepath; do
+        echo -en "${comma}     {\"name\":\"/$filepath\", \"digest\":{\"sha512\":\"$checksum\"}}"
+        if [ -z "$comma" ] ; then
+            comma=',\n'
+        fi
+    done
+   echo -en "\n   ]"
+}
+# --
+
+__slsa_provenance_materials() {
+    local csum="" uri="" repo="" ebuild=""
+
+    local ebuild="${CATEGORY}/${PN}/${PF}.ebuild"
+    if [ -f "$(portageq get_repo_path ${ROOT:-/} coreos)/${ebuild}" ] ; then
+        repo="coreos-overlay"
+    elif [ -f "$(portageq get_repo_path ${ROOT:-/} portage-stable)/${ebuild}" ] ; then
+        repo="portage-stable"
+    else
+        die "SLSA provenance: Unable to detect ebuild repository for package '${ebuild}'"
+    fi
+
+    einfo "    Provenance: recording ebuild material (input) '${repo}/${ebuild}'"
+    echo '   "materials": ['
+
+    # The ebuild. Since "configSource" in "invocation" cannot have more than one (top/level) entry
+    #  we add the ebuild and git repo checksum here, as a material.
+    csum="$(cat "/mnt/host/source/src/scripts/.git/modules/sdk_container/src/third_party/${repo}/ORIG_HEAD")"
+    uri="git+https://github.com/flatcar-linux/${repo}.git/${ebuild}"
+    echo -e "      { \"uri\": \"${uri}\","
+    echo -n "        \"digest\": {\"sha1\":\"${csum}\"} }"
```

Comment:
Couple points:

- I think URI should contain the specific commit this ebuild was taken from, so for instance, it should be in form like `https://raw.githubusercontent.com/flatcar-linux/coreos-overlay/3829e4d0dd8adb5c2af47f543723c3a525b5e743/sys-apps/systemd/systemd-250.3.ebuild`
- ORIG_HEAD contains a previous state of HEAD. It should use HEAD. Also, ORIG_HEAD does not even exist on first update of submodules (I don't have it here).
- It's confusing to record commit checksum of the repo and use URI to an ebuild. Shouldn't it be rather checksum of the ebuild file?

```suggestion
    # The ebuild. Commit checksum is contained within an URI to the ebuild.
    csum="$(cat "/mnt/host/source/src/scripts/.git/modules/sdk_container/src/third_party/${repo}/HEAD")"
    uri="https://raw.githubusercontent.com/flatcar-linux/${repo}/${csum}/${ebuild}"
    # I don't know where exactly we can find the ebuild file in this context.
    esum=$(sha512sum "${PATH_TO_THE_CURRENT_EBUILD_FILE}" | cut -f1 -d' ')
    echo -e "      { \"uri\": \"${uri}\","
    echo -n "        \"digest\": {\"sha512\":\"${esum}\"} }"
```

Reply:
@t-lo any thoughts on the 3rd bullet point?

Reply:
Both is valid in SLSA, the uri determines the semantics of the checksum.
I used the git checksum as it seemed to be more useful for reproducing the build, but Krzesimir's first bullet point addresses this. So yes, if we can find the path to the ebuild and generate its sha512sum that would work for me.

Reply:
> I think URI should contain the specific commit this ebuild was taken from, so for instance, it should be in form like https://raw.githubusercontent.com/flatcar-linux/coreos-overlay/3829e4d0dd8adb5c2af47f543723c3a525b5e743/sys-apps/systemd/systemd-250.3.ebuild

I've switched uri to contain the commit hash according to https://spdx.github.io/spdx-spec/package-information/#77-package-download-location-field.

> ORIG_HEAD contains a previous state of HEAD. It should use HEAD. Also, ORIG_HEAD does not even exist on first update of submodules (I don't have it here).

Fixed that.

> It's confusing to record commit checksum of the repo and use URI to an ebuild. Shouldn't it be rather checksum of the ebuild file?

The csum field now contains the actual file checksum, but the the uri contains the commit hash. I think that covers everything.